quamina.go

```diff
@@ -0,0 +1,94 @@
+package quamina
+
+import (
+	"errors"
+	"fmt"
+)
+
+// Quamina instances provide the public APIs of this pattern-matching library.
+// flattener is responsible for turning the bytes of incoming events into a list of name/value pairs. Each
+//  Quamina instance has its own flattener, because flatteners are stateful and not designed for concurrent use.
+// matcher is the root of the two-level automaton structure containing fieldMatcher and valueMatcher nodes.  Multiple
+//  Quamina instances may have the same matcher value, since it is designed for concurrent operation.
+type Quamina struct {
+	flattener Flattener
+	matcher   matcher
+}
+
+type Option func(q *Quamina) error
+
+func WithMediaType(mediaType string) Option {
+	return func(q *Quamina) error {
+		switch mediaType {
+		case "application/json":
+			q.flattener = newJSONFlattener()
+		default:
+			return fmt.Errorf(`media type "%s" is not supported by Quamina`, mediaType)
+		}
+		return nil
+	}
+}
+func WithFlattener(f Flattener) Option {
+	return func(q *Quamina) error {
+		if f == nil {
+			return errors.New("nil Flattener")
+		}
+		q.flattener = f
+		return nil
+	}
+}
```

Comment:
There's an apply issue with these two options competing for the final outcome and I think those options are either redundant or mutual exclusive - the latter makes it harder with the function options pattern. Let me explain.

Since those are options, we have no control over the order of how a user passes these options. Neither should a user care. However, the following can happen:

`opts := []quamina.Option{WithFlattener(myCustomJSONOne), WithMediaType("application/JSON")}`

In this case, the user would not get the desired `Flattener` but our default `newJSONFlattener`.

My proposal is to drop one of these options, preferably the `MediaType`. Reason:

- The `Flattener` implicitly deals with a specific media type - why do I have to specify one?
- If we drop `WithFlattener`, users lose control over the applied `Flattener` and we'd have to constantly update `WithMediaType` `switch` statement
- If we drop `WithMediaType`, the user gets our default (JSON) or can overwrite the `Flattener` as needed and use a custom (or also JSON) media type based on the implemented `Flattener`

Since `Flattener` is an interface, we don't even have to own these in our repo. A user can "bring" it and does not have to care about the semantics of also calling `WithMediaType`.

Not sure if I'm understanding the rational behind `WithMediaType` well. Looks like it's purely for `Flattener` selection which complicates the functional option pattern a lot. For me the media type behavior is implicit and the `Flattener` would throw an error if it gets a mismatched event anyways.

Reply:
I think on this one we'll have to disagree.  The Internet has a well-established, well-debugged, broadly implemented standard for identifying data formats, so if we're going to identify data formats, we should use it. I think we still need WithFlattener for custom Flatteners that aren't associated with a well-known media type or offer some nonstandard custom feature. I agree that specifying both WithFlattener() and WithMediaType() doesn't make sense, so I'll put a check in the code to throw an error. Also doesn't make sense to specify both WithPatternDeletion(true) and WithPatternDeletion(false).